contracts/SpokePoolV3Periphery.sol

```diff
@@ -11,28 +11,71 @@ import { V3SpokePoolInterface } from "./interfaces/V3SpokePoolInterface.sol";
 import { IERC20Auth } from "./external/interfaces/IERC20Auth.sol";
 import { WETH9Interface } from "./external/interfaces/WETH9Interface.sol";
 
+/**
+ * @title SpokePoolProxy
+ * @notice User should only interact with SpokePool via the SpokePoolV3Periphery contract through this
+ * contract. This is purposefully a simple passthrough contract so that the user only approves this contract
+ * to pull its assets while the SpokePoolV3Periphery contract can be used to call any calldata on any exchange
+ * that the user wants to. By separating the contract that gets approved from the contract that executes arbitrary
+ * calldata, the SpokePoolPeriphery does not need to validate the calldata that gets executed.
+ * @dev If this proxy didn't exist and users instead approved and interacted directly with the SpokePoolV3Periphery
+ * then users would run the unneccessary risk that another user could instruct the Periphery contract to steal
+ * any approved tokens that the user had left outstanding.
+ */
+contract SpokePoolProxy is Lockable {
+    using SafeERC20 for IERC20;
+
+    error LeftoverInputTokens();
+
+    // The SpokePoolPeriphery should be deterministically deployed at the same address across all networks,
+    // so this contract should also be able to be deterministically deployed at the same address across all networks
+    // since the periphery address is the only constructor argument.
+    address public immutable SPOKE_POOL_PERIPHERY;
+
+    constructor(address _spokePoolPeriphery) {
+        SPOKE_POOL_PERIPHERY = _spokePoolPeriphery;
+    }
+
+    /**
+     * @notice Caller must insure that exactly `inputAmount` of `inputToken` is used in the subsequent call to
+     * the SpokePoolPeriphery contract. All of the periphery functions have an easily identifiable input
+     * amount and input token so this should be easy to verify. Any leftover tokens would be locked in this contract
+     * and could be used by ANYONE in a subsequent call to the SpokePoolPeriphery contract. So, this function
+     * attempts to protect the user from locking tokens by reverting if not exactly `inputAmount` of `inputToken`
+     * is used in the periphery contract call.
+     */
+    function callSpokePoolPeriphery(
+        address inputToken,
+        uint256 inputAmount,
+        bytes memory peripheryFunctionCalldata
+    ) external payable nonReentrant {
+        uint256 balanceBefore = IERC20(inputToken).balanceOf(address(this));
+        IERC20(inputToken).safeTransferFrom(msg.sender, address(this), inputAmount);
+        IERC20(inputToken).forceApprove(SPOKE_POOL_PERIPHERY, inputAmount);
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, bytes memory result) = SPOKE_POOL_PERIPHERY.call{ value: msg.value }(peripheryFunctionCalldata);
```

Comment:
I think we only want to use this proxy contract if we are calling `swapAndBridge` on the periphery, so should we just make this function a wrapper for `swapAndBridge` specifically?

Reply:
are you saying that because the other flows wouldn't require an `approve`?

If so, do you think the user of this Periphery contract would prefer to just go through a single contract address (i.e.g the proxy)? Or do you think the moderate gas cost increase is worth avoiding and switching on the address to call?

Reply:
https://github.com/across-protocol/contracts/pull/777/commits/95116666524b72d84901c6e0e083d306ecf63700

Reply:
Ah, I do see the value here. So for the permit/2/transferWithAuth scenarios, the `transferFrom` would be transferring 0 tokens to the contract, and the benefit would be for integrators (and our FE) to not need to juggle two addresses for a special case of swap and bridging, right? @dohaki, would you prefer just one address here? I personally think the extra gas is not awful, and maybe you could then just conditionally approve the peripery conract /transfer from the user if the amount to transfer is not zero.

Reply:
the downside of generalizing this function is we would probably want to add some protection to the user that they don't leave any tokens in this contract:

```solidity
uint balanceBefore = x;
spokePoolPeriphery.call();
require(balanceAfter == balanceBefore, "user left tokens on proxy");
```

which would add gas.